### PR TITLE
Automate the testing of popular blocks in current and edge versions of GB

### LINKED ISSUE #45209: [Gutenberg WPCOM e2e tests i0] Automate steps 1 to 7 (Tracking Issue)
**Project:** pc9Uba-4N-p2
**Task:** Automate steps 1 to 7 from the Good Mountain Testing Checklist.

### Status

After the first exploration/PoC phase, I came up with a [PR](https://github.com/Automattic/wp-calypso/pull/45048) that gets close to automating steps 1 to 7 while testing across the most popular themes and non-edge/edge sites, but there are a few aspects missing, which leads us to our TODO list, below.

### TODO

- [X] Initial groundwork / exploration [task](https://github.com/Automattic/wp-calypso/pull/45048)
- [x] Fill all the popular blocks with the necessary minimum data for them to appear in the frontend and make sure each of them is displayed (here's an [example](https://href.li/?https://github.com/Automattic/wp-calypso/blob/master/test/e2e/specs/wp-calypso-gutenberg-coblocks-spec.js#L69) of how the CoBlocks e2e tests do it) (@fullofcaffeine)
- [ ] ~DRY the frontend selector definition for each block class by deriving them at runtime, see: https://github.com/Automattic/wp-calypso/pull/45794/files#r495125663~ Moved to a task, [here](https://github.com/orgs/Automattic/projects/185#card-46874188), since it's not really required for this issue to be closed (it's an improvement).
- [ ] ~For more elaborate/interactive blocks, mimick a user using it the frontend. It might be as simple as clicking, say, the tile-gallery, and navigating through the images while making sure nothing broke. For each block, check if the codebase where it's developed already has a testing infrastructure/tests for it, and if so, make sure we can change the GB version to the one we want, and add the test there (the tests might already exist, just make sure it runs with a specific version of GB). Skip coblocks for now.~ This will be part of the next iteration and we will make sure that these tests are upstream, whenever possible.

### PR DESCRIPTION
**Pull request by fullofcaffeine**
#### Changes proposed in this Pull Request

* Automates steps 1 to 7 of the Good Mountain testing checklist but does not focus on testing individual blocks (this will come in subsequent PR(s));
* Look for blocks on the page and also look for block errors on the page;
* Takes screenshots from edge and non-edge editor and preview.
* Do that for edge/non-edge sites across the most popular themes
* Adds a new test account and several new test sites (the account credentials have been added to the encrypted config file)

Repeat that for the most popular themes in wpcom.

#### Testing instructions

1. Setup your env to run Calypso E2E tests (search for "Automated end-to-end Testing") in the FG;
1. Run with `./node_modules/.bin/mocha specs/wp-calypso-gutenberg-upgrade-spec.js`*[0]
1. Wait for the tests to finish. It might take a good while since it tests everything for each test site
1. Make sure it didn't err, if it did, it might not be an actual failure, try running that specific step again. If it continues failing, let us know
1. Look inside the screenshot*[1] folder for the screenshots. There should be a directory per test site name. You can find the list of test sites in the [test file](https://github.com/Automattic/wp-calypso/pull/45048/files#diff-87bdd0c59221389b9b5d297efbdb1b31R117) itself. They are all part of the same test account called `gutenbergUpgradeUser`. You can verify it (alongside the credentials) by decrypting the config file in this PR and looking for it in your `config/local-decrypted.json` file.

*[0] _Make sure to cd into `test/e2e` first._
*[1] `test/e2e/screenshots/`

#### TODO

- [x] Figure out if setting up blocks in a wpcom site then copying the block markup to another renders the blocks correctly in the new site (it looks like it does! This means we can postpone for now the idea of switching gutenberg or applying stickers using a query param)
- [x] Populate tiled-gallery with sample images
- [x] Take screenshots of the editor and preview in the non-edge site
- [x] Take screenshots of the editor and preview in the edge site, after loading the blocks markup from non-edge
- [x] Research how to set it up as a scheduled test and to run it using the "Needs e2e testing Gutenberg edge" tag (we'll just let it run as part of the current e2e pipeline)

### CODE REVIEW

File: test/e2e/lib/gutenberg/blocks/contact-info-block-component.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+/**
+ * External dependencies
```
Comment (fullofcaffeine) [inline test/e2e/lib/gutenberg/blocks/contact-info-block-component.js:2]: Left those comments for now as I'm not sure if I'll change these files soon, but will remove in a cleanup before taking this PR out of the draft state.

File: test/e2e/lib/gutenberg/gutenberg-editor-component.js [hunk 2]
```diff
@@ -145,16 +156,199 @@ export default class GutenbergEditorComponent extends AsyncBaseContainer {
...
+			codeEditor
+		);
+	}
+
+	async pasteBlocksCode() {
+		const codeEditor = await this.switchToCodeView();
+		return codeEditor.sendKeys( Key.CONTROL + 'v' );
+	}
+
+	async setBlocksCode( code ) {
```
Comment (fullofcaffeine) [inline test/e2e/lib/gutenberg/gutenberg-editor-component.js:211]: This is not used anymore and will be removed.

File: test/e2e/lib/gutenberg/gutenberg-editor-component.js [hunk 3]
```diff
@@ -145,16 +156,199 @@ export default class GutenbergEditorComponent extends AsyncBaseContainer {
...
+		);
+
+		// Now click to switch to the block editor
+		await driverHelper.clickWhenClickable(
+			this.driver,
+			By.xpath( "//div[@aria-label='More tools & options']/div[2]/div[2]/button[1]" )
+		);
+	}
+
+	async getBlocksCode() {
```
Comment (fullofcaffeine) [inline test/e2e/lib/gutenberg/gutenberg-editor-component.js:192]: This is not used anymore and will be removed

File: test/e2e/specs/wp-calypso-gutenberg-upgrade-spec.js [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+				const contactFormDisplayedInEditor = await gEditorComponent.contactFormDisplayedInEditor();
+				assert.strictEqual(
+					contactFormDisplayedInEditor,
+					true,
+					'The contact form is not displayed in the editor'
+				);
+
+				// jetpack/layout-grid
+				await gEditorComponent.insertLayoutGrid();
+				const layoutGridDisplayedInEditor = await gEditorComponent.layoutGridDisplayedInEditor();
```
Comment (fullofcaffeine) [inline test/e2e/specs/wp-calypso-gutenberg-upgrade-spec.js:81]: There's quite a bit of repetition here and I don't like it. I'll see how to dry it up later.

File: test/e2e/lib/gutenberg/blocks/blog-posts-block-component.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+/**
+ * Internal dependencies
+ */
+import GutenbergBlockComponent from './gutenberg-block-component';
+
+class BlogPostsBlockComponent extends GutenbergBlockComponent {}
+
+BlogPostsBlockComponent.blockTitle = 'Blog Posts';
+BlogPostsBlockComponent.blockName = 'a8c/blog-posts';
```
Comment (sirreal) [inline test/e2e/lib/gutenberg/blocks/blog-posts-block-component.js:9]: This seems more idiomatic:

```suggestion
class BlogPostsBlockComponent extends GutenbergBlockComponent {
	static blockTitle = 'Blog Posts';
	static blockName = 'a8c/blog-posts';
}
```

File: test/e2e/lib/gutenberg/blocks/index.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+// TODO some blocks are exported as default and some not... maybe change all of them to export as default?
```
Comment (sirreal) [inline test/e2e/lib/gutenberg/blocks/index.js:1]: Consistency is likely the best approach. Maybe in another PR we change them all to use one or the other.

Slight preference from me for named exports here.

File: test/e2e/lib/gutenberg/blocks/image-block-component.js [hunk 7]
```diff
@@ -10,10 +10,6 @@ import * as driverHelper from '../../driver-helper';
 import GutenbergBlockComponent from './gutenberg-block-component';
 
 export class ImageBlockComponent extends GutenbergBlockComponent {
-	constructor( driver, blockID ) {
```
Comment (sirreal) [inline test/e2e/lib/gutenberg/blocks/image-block-component.js:13]: This cleanup is nice, but it distracts from the focus on this PR. There's a lot of cleanup like this that we can probably do in a few tiny, focused PRs.

File: test/e2e/lib/gutenberg/blocks/tiled-gallery-block-component.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+
+/**
+ * Internal dependencies
+ */
+import GutenbergBlockComponent from './gutenberg-block-component';
+
+class TiledGalleryBlockComponent extends GutenbergBlockComponent {
+	async uploadImages( filesDetails ) {
+		const fileInput = this.driver.findElement(
+			By.xpath( `//*[@id="${ this.blockID.slice( 1 ) }"]/div/div/div[3]/div[2]/input` )
```
Comment (sirreal) [inline test/e2e/lib/gutenberg/blocks/tiled-gallery-block-component.js:14]: 😅  Oh boy… what's this targeting? Something in the media placeholder? Maybe a dedicated helper would be good?

**Review by sirreal — COMMENTED** [review]
Nice start here, this is probably ready for more final reviews? Shall we swap the in progress label?

File: test/e2e/lib/gutenberg/blocks/blog-posts-block-component.js (hunk 5 above)
Comment (fullofcaffeine) [reply test/e2e/lib/gutenberg/blocks/blog-posts-block-component.js:9]: Oh, that was the first thing I tried. Unfortunately, I got a syntax error. The tests don't go through a transpiler, so it fallbacks to what node supports, and AFAIK, it still doesn't support `static` class properties.

Comment (fullofcaffeine) [reply test/e2e/lib/gutenberg/blocks/blog-posts-block-component.js:9]: Using node `12.16.2`:

```
 ./node_modules/.bin/mocha specs/wp-calypso-gutenberg-upgrade-spec.js
file:///home/fullofcaffeine/workspace/code/wp-calypso/test/e2e/lib/gutenberg/blocks/youtube-block-component.js:7
        static blockTitle = 'Foo';
                   ^

SyntaxError: Invalid or unexpected token
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1176:10)
```

File: test/e2e/lib/gutenberg/blocks/image-block-component.js (hunk 7 above)
Comment (fullofcaffeine) [reply test/e2e/lib/gutenberg/blocks/image-block-component.js:13]: Copy that 👍🏻 , I'll keep that in mind.

File: test/e2e/lib/gutenberg/blocks/blog-posts-block-component.js (hunk 5 above)
Comment (fullofcaffeine) [reply test/e2e/lib/gutenberg/blocks/blog-posts-block-component.js:9]: Oh, but it does support static getters:

```javascript
class YoutubeBlockComponent extends GutenbergBlockComponent {
	static get blockTitle() {
		return 'foo ';
	}
}
```

Comment (fullofcaffeine) [reply test/e2e/lib/gutenberg/blocks/blog-posts-block-component.js:9]: I think I'll use static getters for now.

Comment (sirreal) [reply test/e2e/lib/gutenberg/blocks/blog-posts-block-component.js:9]: I think it's an issue with how this is configured. I believe mocha is using its own parser. In node 12.18.0 it works:

```sh
> class A {}
undefined
> class B { a='a';}
undefined
> A.a
undefined
> B.a
undefined
> new B().a
'a'
> A.a='a'
'a'
> A.a
'a'
> class C { static a = 'a' }
undefined
> C.a
'a'
```

Let's just leave it for this PR. It's not really relevant.

Comment (fullofcaffeine) [reply test/e2e/lib/gutenberg/blocks/blog-posts-block-component.js:9]: Yeah, I tested it with vanilla node and saw it worked. I tried to figure out what as the culprit in mocha but couldn't find and left it aside. I'm using static getters for now.

I'd be interested to know what is the culprit here. I saw the mocha js binary and AFAIK it uses node as well. That's pretty confusing.

File: test/e2e/lib/gutenberg/blocks/tiled-gallery-block-component.js (hunk 8 above)
Comment (fullofcaffeine) [reply test/e2e/lib/gutenberg/blocks/tiled-gallery-block-component.js]: This selects the hidden file input inside the tile gallery. I've moved this to its own instance field with a more descriptive name. Let me know if you have anything else in mind.

**Comment by fullofcaffeine** [comment]
@Automattic/good-mountain @Automattic/e2e-test-reviewers Please have a look and let me know what you think. 

Please do bear in mind that it might seem incomplete - the test data for most of the blocks are not being filled yet (only for two of them) - but that was a deliberate decision taken to allow us to finish the initial groundwork/exploration task earlier instead of having a big/long-running PR (this version already provides some value and we can use it to help us test the next GB version).

Also, the assertion for invalid blocks has been commented out for now due to an [issue](https://github.com/Automattic/jetpack/issues/16514) in the `jetpack/tiled-gallery`. I'm not sure how to proceed here and could use some input. I see a couple of alternatives: 1) We can ignore this step for the `jetpack/tiled-gallery` for now; 2) We can leave it commented out until the issue is fixed; 3) We can remove tiled-gallery from the checks?

File: test/e2e/lib/gutenberg/blocks/blog-posts-block-component.js (hunk 5 above)
Comment (sirreal) [reply test/e2e/lib/gutenberg/blocks/blog-posts-block-component.js]: Fix for this in https://github.com/Automattic/wp-calypso/pull/45268.

**Comment by sirreal** [comment]
Rebased with #45268 merged.

**Review by sirreal — APPROVED** [review]
Nice work here! 🚢

**Comment by sirreal** [comment]
I don't see the screenshots artifacts in https://app.circleci.com/pipelines/github/Automattic/wp-e2e-tests-for-branches/36242/workflows/a894971d-692c-4e0e-9951-884404b1774c/jobs/97521/artifacts - is the plan to store them there?

We can introduce that in a follow-up, it should be a small change to the CircleCI configuration.

**Comment by fullofcaffeine** [comment]
Hmm, thanks for pointing that out! I just assumed that by saving them to the screenshots folder, everything else would fall into place, but there might be something in CircleCI as you said. I'll take note of that.

**Pull request merged**